### PR TITLE
Add index command

### LINKED ISSUE #639: Add index subcommand to summarize sequence content
**Context**

While discussing a new standardized interface for reading and writing sequences in augur, we realized that the only time we need random access to sequence inputs (e.g., through BioPython's `SeqIO.index`) is when `augur filter` needs to assess the nucleotide content of sequences.

Augur's filter steps could be substantially sped up if they only needed to  inspect metadata tables instead of parsing large sequence files. Specifically, if we could calculate sequence statistics ahead of time and merge than with existing metadata, we could filter by columns in a table without needing to read sequences at each filter step.

**Description**   

@tsibley recommended the addition of an `augur index` command (or something similarly named) that could calculate the sequence composition statistics we need one time.

**Examples**

```bash
# This should work with standard FASTA files.
augur index --sequences data/sequences.fasta --output results/sequence_composition.tsv

# This should also work with compressed FASTA files.
augur index --sequences data/sequences.fasta.gz --output results/sequence_composition.tsv
```

**Possible solution**

Similar existing tools include `samtools faidx` (provides a table of sequence lengths by name) or `seqtk comp` from [Heng Li's seqtk toolkit](https://github.com/lh3/seqtk). For example, the following command might provide exactly what we need:

```bash
seqtk comp data/sequences.fasta > results/sequence_composition.tsv
```

The output looks like this for ncov sequences where the output columns are `chr, length, #A, #C, #G, #T, #2, #3, #4, #CpG, #tv, #ts, #CpG-ts` (`#4` [may represent the number of N bases](https://github.com/lh3/seqtk/issues/47)):

```
Wuhan/IVDC-HB-01/2019   29891   8942    5492    5863    9594    0       0       0       878     0       0       0
Wuhan/IVDC-HB-04/2020   29896   8947    5490    5863    9596    0       0       0       878     0       0       0
Wuhan/IVDC-HB-05/2019   29891   8944    5492    5861    9594    0       0       0       874     0       0       0
```

seqtk comp takes ~1 minute to write out the composition of a 7.4 GB file and can also ingest compressed FASTA files. Augur could wrap these fast tools in its own interface to abstract the original program interface (e.g., to provide column headings with clearer names).

### PR DESCRIPTION
**Pull request by benjaminotter**
**Description of proposed changes**    
Add an `augur index` command that returns sequence statistics, namely the nucleotide content, and writes the output to a .tsv file. This data can then be used during `augur filter` when access to sequence statistics is needed instead of parsing through the input sequences.
`augur index` takes a fasta file as input, assesses the nucleotide content of each sequence and writes the counts to a .tsv file with the columns `sequence_name, length, A, C, G, T, N, other_IUPAC`.

Add `index` to command strings in `__init__.py`

**Example use**
`augur index --sequences data/sequences.fasta --output results/sequence_composition.tsv`

**Related issue(s)**  
Fixes #639

### CODE REVIEW

File: augur/index.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    '''
+    #read in files
+    try:
+        seqs = SeqIO.parse(args.sequences, 'fasta')
+    except ValueError as error:
+        print("ERROR: Problem reading in {}:".format(args.sequences))
+        print(error)
+        return 1
+
+    good_chars = {'a', 'c', 'g', 't', 'n', '-', 'r', 'y', 's', 'w', 'k', 'm', 'd', 'h', 'b', 'v','?'}
```
Comment (rneher) [inline augur/index.py:25]: Why don't you construct `good_chars` from the `values`. Then one wouldn't need to keep to separate sets up-to-date.

File: augur/index.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    with open(args.output, 'wt') as out_file:
+        tsv_writer = csv.writer(out_file, delimiter = '\t')
+
+        #write header
+        header = ['sequence_name', 'length']+labels
+        tsv_writer.writerow(header)
+
+        for record in seqs:
+            #check if sequence contains non nucleotide characters
+            if len(set(str(record.seq.lower())).difference(good_chars))==0:
```
Comment (rneher) [inline augur/index.py:48]: I wouldn't skip them. Why don't you calculate the number of invalid characters and report them separately? The `index` command will probably be used to get an overview of sequences, so having stats on invalid characters would be useful.

**Review by rneher — CHANGES_REQUESTED** [review]

**Comment by rneher** [comment]
@huddlej, @benjaminotter made a first stab at an index command. Would be great to get your feedback and how we would want to integrate this info in `filter`

**Comment by huddlej** [comment]
Thanks (and hello!), @benjaminotter! This is a great start. I'll post some specific feedback Thursday before 12pm PST and push some example code for how the output of this new command will feed into the augur filter command.

**Review by huddlej — COMMENTED** [review]
@benjaminotter, thank you again for taking the initiative to work on this new subcommand. The code itself is fast (processes 260,000 ncov genomes in 4.5 minutes) and easy to read. Most of my comments in this review are minor points that relate to Augur-specific design preferences that you wouldn't have been exposed to before.

The idea of an `index` command is only a couple of weeks old and we haven't discussed its interface much within the Nextstrain group. The main goal of the sequence index file is to avoid random access reading of sequences during `augur filter` by precalculating necessary data about the sequences. Your code has helped me implement a prototype of how this sequence index can be used in `augur filter` to address this random access issue. I will push a couple of commits soon to this branch that include this prototype.

Two major points of feedback on the current code:

1. We need to count all invalid nucleotides, since these are used by `augur filter` to exclude sequences. We should also count ambiguous or gap characters like `?` and `-`, since these are considered "good characters" by augur filter. I imagine the final output will include the following columns:

  - strain
  - length
  - A
  - C
  - G
  - T
  - N
  - other_IUPAC
  - `?`
  - `-`
  - invalid_nucleotides

The sum of the counts for columns from "A" through "invalid_nucleotides" should equal the value in the "length" column. My inline comments include recommendations about how to avoid repeating lists of valid nucleotides by using existing definitions in Augur.

2. To provide a backward-compatible `augur filter` interface that does not require a sequence index, we would need a way to calculate this index on demand from another Python module. If we choose to go this direction, we'll need to refactor the bulk of the logic from `run` into its own function in the `index.py` module that can receive paths to input sequences and an output index file and be called from the `filter.py` module. I mocked up an example of how this function would be called in the prototype filter code I'm about to push.

Alternately, we can choose to change the `augur filter` API to make the `--sequence-index` a required input. Even if we don't go this way, refactoring the index logic into its own function(s) will make it easier to read and test.

**Comment by huddlej** [comment]
@benjaminotter and @rneher, commit 49fc339 adds a prototype of the new augur filter interface and stubs out two new functions in `index.py` that I propose as a way to support on-the-fly indexing from `augur filter` (for backwards-compatibility with workflows that do not precalculate the index) and as a way to write tests for the core logic of the index command.

The commit message provides more details.

I tested this prototype with the [Zika tutorial](https://github.com/nextstrain/zika-tutorial) sequences like so:

```
# Build the index file.
augur index \
  --sequences data/sequences.fasta \
  --output results/sequence_index.tsv

# Filter with the index file. This should work.
augur filter \
  --sequences data/sequences.fasta \
  --metadata data/metadata.tsv \
  --sequence-index results/sequence_index.tsv \
  --min-length 8000 \
  --min-date 2016-01-01 \
  --output results/filtered.fasta

# Filter without the index file. This won't work yet.
augur filter \
  --sequences data/sequences.fasta \
  --metadata data/metadata.tsv \
  --min-length 8000 \
  --min-date 2016-01-01 \
  --output results/filtered.fasta
```

**Comment by rneher** [comment]
Thanks, @benjaminotter! this looks good. I just check how fast the `SimpleFastaParser` is compared to `SeqIO.parse`. about 20-30% fasta (3:30 vs 4:50). Not worth the trouble.

**Comment by benjaminotter** [comment]
@huddlej, commit [`f5ae462`](https://github.com/nextstrain/augur/pull/651/commits/f5ae462b0351cb800cf2d64240091559f44ee560) makes `augur filter` work with on the fly indexing. 
I tested it with the Zika Tutorial as you mentioned in your comment above:

> ```
> # Build the index file.
> augur index \
>   --sequences data/sequences.fasta \
>   --output results/sequence_index.tsv
> 
> # Filter with the index file. This should work.
> augur filter \
>   --sequences data/sequences.fasta \
>   --metadata data/metadata.tsv \
>   --sequence-index results/sequence_index.tsv \
>   --min-length 8000 \
>   --min-date 2016-01-01 \
>   --output results/filtered.fasta
> 
> # Filter without the index file. This won't work yet.
> augur filter \
>   --sequences data/sequences.fasta \
>   --metadata data/metadata.tsv \
>   --min-length 8000 \
>   --min-date 2016-01-01 \
>   --output results/filtered.fasta
> ```

Filter without index file now calls the new functions in `index.py` and the output is the same as prebuilding the index file and using it as input when calling `augur filter`.

**Comment by tsibley** [comment]
> processes 260,000 ncov genomes in 4.5 minutes

This seemed quite slow to me on the face of it. Reading the code, I was struck by the multiple invocations of `seq.count()` which seemingly must mean multiple passes over each entire sequence. I thought it might be faster to iterate over the sequence just once ourselves recording what we saw, and then bin those observed counts into the value sets of interest. Turns out this isn't quicker though, as multiple passes of the C implementation of `str.count()` is still faster than a single pass Python implementation. I bet we'd benefit from a single-pass C implementation, but I'm not going to pursue that unless other folks think it's worthwhile.

In the process of this, I corrected the existing doctests and added some argument validation to the internal function to avoid returning incorrect results. Pushed commits for each.

**Comment by huddlej** [comment]
Thank you for the code review, tests, and validation, @tsibley!

> I bet we'd benefit from a single-pass C implementation, but I'm not going to pursue that unless other folks think it's worthwhile.

My first inclination when writing [the original issue](https://github.com/nextstrain/augur/issues/639) was to make a light wrapper around `seqtk comp` and clean up its output a bit. It's written in C and generates roughly the same output in 46 seconds with:

```
seqtk comp data/sequences.fasta > sequence_index.tsv
```

Now that we have this implementation, though, I'm happy with the flexibility and control it provides us without the complexity of another dependency or new C code in the repo. If we get to the point where it's too slow, we could revisit the single-pass C approach though.

**Comment by rneher** [comment]
@huddlej are we ok merging this? There are definitely some perf improvements that one could do, but it is not our bottleneck. 


FWIW:
```python
In [8]: timeit s.count('A')
117 µs ± 1.61 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [9]: timeit Counter(list(s))
2.39 ms ± 22.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
so counting 20 times is about as fast as using `Counter` to count all at once....

**Comment by huddlej** [comment]
@rneher I just want to test this with a couple of different builds to make sure it doesn't break anything we missed in our tests. I'll work on this on Monday.

We should also add a warning message to augur filter when the index gets created on the fly, so people know what command to run to pre-build the index. I'd like to add an example to docs, too, at least in the Zika tutorial if not the ncov tutorial.

After those minor checks and docs, I think we're safe to merge.

**Comment by rneher** [comment]
I ran into one issue here. you often want to filter an existing data set multiple times. if no index is provided, this will attempt to recreate the index and overwrite the file. We should only filter when we actually need `min_length` and `non-ATCG` counts and make sure we don't attempt to overwrite the file.

**Comment by huddlej** [comment]
> if no index is provided, this will attempt to recreate the index and overwrite the file.

I see that now. So, it's possible that one filter command could overwrite an existing filter command process's index with an empty file before the original process can read the index. I'm not as worried about the wasted time spent recreating the index (since we do this already in `master`), but the auto-generated index should have a unique temp filename instead of the current name. I just pushed a commit with this change.

> We should only filter when we actually need min_length and non-ATCG counts and make sure we don't attempt to overwrite the file.

We currently use the list of strain names from the sequence records to define our sequences to keep prior to filtering. If we source this list from the metadata instead, we could skip auto-building the index when it isn't needed. Since [the first step of filtering is to ignore all sequences that don't have metadata](https://github.com/nextstrain/augur/blob/8d69df645b2bfce8778861402ad5cf3a104d8fb3/augur/filter.py#L174), this wouldn't be a big change.

My only concern with using metadata as the source of truth is that the warnings from augur filter about sequences without metadata could be the only time people realize they are missing metadata. Maybe we could address this more systematically by adding sequence/metadata file validation to the `augur validate` command or somewhere similarly reasonable.

Also, heads up that I rebased this branch onto `master` since things have diverged a bit and tests are easier to run now.

**Comment by rneher** [comment]
the other issue with not checking upfront that sequences exist is in the case where there is metadata but no sequence. subsampling then operates on the wrong set of strains.

**Comment by huddlej** [comment]
Ok, we can leave this as it is for this PR. I've tested index with the Zika tutorial and I'm testing with the ncov build next.

File: augur/filter.py [hunk 3]
```diff
@@ -141,16 +146,41 @@ def run(args):
...
+            sep="\t"
+        )
+
+        # Remove temporary index file, if it exists.
+        if index_is_autogenerated:
+            os.unlink(sequence_index_path)
+
+        # Calculate summary statistics needed for filtering.
+        sequence_index["ACGT"] = sequence_index.loc[:, ["A", "C", "G", "T"]].sum(axis=1)
+        sequence_index["valid_nucleotides"] = sequence_index.loc[:, ["ACGT", "N", "other_IUPAC"]].sum(axis=1)
```
Comment (huddlej) [inline augur/filter.py:182]: Valid nucleotides should include "?" according to the original implementation of "non-nucleotide".

File: tests/functional/filter.t [hunk 4]
```diff
@@ -52,6 +55,7 @@ Using the default probabilistic subsampling, should work the same as the previou
 
   $ ${AUGUR} filter \
   >  --sequences filter/sequences.fasta \
+  >  --sequence-index filter/sequences.fasta.idx \
```
Comment (huddlej) [inline tests/functional/filter.t:58]: These test files should be called `sequence_index.tsv` for consistency across documentation and workflows.

**Comment by huddlej** [comment]
@rneher I've tested this now with Zika tutorial and a full global ncov build without any issues.

One potential annoyance in the ncov workflow is that [the simplest place to create the sequence index is at the start of the workflow](https://github.com/nextstrain/ncov/pull/552/files#diff-428619734d758efae03a32c735ef532b2c276cf600a953719ec75c6f62c67e60) from the initial `sequences.fasta`. When you try to start the workflow from the S3 filtered FASTA, though, this setup requires downloading both the original sequences FASTA and the filtered file. We could address this with something like a secondary `filtered_sequence_index.tsv`, though.

I don't think the annoyance is a blocker for the interface proposed here. If you agree, I will merge and release this as is.

**Comment by rneher** [comment]
I agree. this is a little annoying. I actually moved the `filter` rule in my `ncov` branch to the very end and removed all other filters:

https://github.com/nextstrain/ncov/blob/nextalign/workflow/snakemake_rules/main_workflow.smk#L274

We were initially worried about messed up sequences complicating alignments, but if we reference align only, this is no longer an issue.

**Comment by huddlej** [comment]
@benjaminotter Thank you again for writing this new index command. Your code is now part of [Augur version 11.1.0](https://github.com/nextstrain/augur/blob/master/CHANGES.md#1110-12-february-2021). Good luck with the rest of your undergrad studies and don't hesitate to reach out if you'd like to contribute to Nextstrain again!

**Comment by benjaminotter** [comment]
@huddlej Thank you very much! It was a pleasure to contribute to Augur and I'll get in touch if there is the possibility for working on some project again.

**Pull request merged**